File: packages/ailog_flutter/ios/Classes/AilogFlutterPlugin.swift
import Flutter
import UIKit

/// Flutter plugin entry point. Wires the `dev.ailog/flutter` MethodChannel
/// in both directions — see `Ailog.swift` for the native→Dart direction
/// and `AilogNativeBridge` (Dart) for the Dart-side handler.
///
/// Also installs (once per process) `NSSetUncaughtExceptionHandler`,
/// chaining to whatever handler was previously installed, so existing
/// crash reporters (Crashlytics, Sentry, ...) and default crash behavior
/// are unaffected.
///
/// **Coverage limitation** (read before relying on this for crash
/// reporting): `NSSetUncaughtExceptionHandler` only observes Objective-C
/// level `NSException`s. It does **not** see:
///  - Swift runtime traps — force-unwrapping `nil`, Swift array
///    out-of-bounds, integer overflow, `fatalError()` — these abort the
///    process directly via a signal, bypassing `NSException` entirely.
///  - Any other signal-raised crash (`SIGSEGV`, `SIGBUS`, `SIGILL`, ...).
///
/// Catching those requires an async-signal-safe C signal handler (no
/// Foundation, no Swift runtime, no allocation inside the handler), which
/// this plugin does not implement — that's a meaningfully larger and
/// riskier undertaking than the MethodChannel bridge this package is
/// centered on. For full native crash coverage, pair this with a
/// dedicated crash reporter and treat `Ailog`'s crash-time write as a
/// best-effort supplement, not a replacement.
public class AilogFlutterPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?

    private static let installLock = NSLock()
    private static var crashHandlerInstalled = false
    private static var configuredLogFilePath: String?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "dev.ailog/flutter", binaryMessenger: registrar.messenger())
        let instance = AilogFlutterPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)

        Ailog.attachChannel(channel)
        installCrashHandlerOnce()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            if let args = call.arguments as? [String: Any], let path = args["logFilePath"] as? String {
                AilogFlutterPlugin.configuredLogFilePath = path
            }
            result(nil)
        case "emitTestLog":
            Ailog.info(
                "test message from native (iOS \(UIDevice.current.systemVersion))",
                context: ["origin": "emitTestLog"],
                tags: ["native", "smoke-test"]
            )
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private static func installCrashHandlerOnce() {
        installLock.lock()
        defer { installLock.unlock() }
        if crashHandlerInstalled { return }
        crashHandlerInstalled = true

        let previousHandler = NSGetUncaughtExceptionHandler()
        NSSetUncaughtExceptionHandler { exception in
            AilogJsonlWriter.write(
                path: AilogFlutterPlugin.configuredLogFilePath,
                level: "fatal",
                message: exception.reason ?? exception.name.rawValue,
                loggerName: Ailog.defaultLoggerName,
                error: NativeError(
                    type: exception.name.rawValue,
                    message: exception.reason ?? exception.name.rawValue,
                    frames: exception.callStackSymbols
                )
            )
            previousHandler?(exception)
        }
    }
}

File: packages/ailog_flutter/android/src/main/kotlin/dev/ailog/ailog_flutter/AilogFlutterPlugin.kt
package dev.ailog.ailog_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlin.system.exitProcess

/**
 * Flutter plugin entry point. Wires the `dev.ailog/flutter` MethodChannel
 * in both directions:
 *
 *  - **Dart → native**: handles `configure` (stores the JSONL file path
 *    used by the crash-time fallback writer) and `emitTestLog` (a
 *    round-trip smoke test — see the example app).
 *  - **Native → Dart**: [Ailog]'s `trace`/`info`/`warn`/`error`/`fatal`
 *    calls are sent out over this same channel and land in
 *    `AilogNativeBridge` on the Dart side.
 *
 * Also installs (once per process) an uncaught exception handler that
 * writes crash events directly to the configured JSONL file via
 * [AilogJsonlWriter], for the case where the exception unwinds after the
 * Flutter engine is no longer available to relay it over the channel. That
 * handler always chains to whatever handler was previously installed, so
 * normal Android crash reporting (Play Console, Crashlytics, ...) and
 * process-termination behavior are unaffected.
 */
class AilogFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private var channel: MethodChannel? = null

    companion object {
        private var crashHandlerInstalled = false
        private var configuredLogFilePath: String? = null
        private val installLock = Any()

        private fun installCrashHandlerOnce() {
            synchronized(installLock) {
                if (crashHandlerInstalled) return
                crashHandlerInstalled = true

                val previous = Thread.getDefaultUncaughtExceptionHandler()
                Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
                    try {
                        val frames =
                            throwable.stackTrace.take(20).map { element ->
                                "${element.className}.${element.methodName}(${element.fileName}:${element.lineNumber})"
                            }
                        AilogJsonlWriter.write(
                            path = configuredLogFilePath,
                            level = "fatal",
                            message = throwable.message ?: throwable.javaClass.name,
                            loggerName = Ailog.defaultLoggerName,
                            error =
                                NativeError(
                                    type = throwable.javaClass.name,
                                    message = throwable.message ?: throwable.javaClass.name,
                                    frames = frames,
                                ),
                        )
                    } catch (_: Throwable) {
                        // Never let crash logging itself mask the original crash.
                    }

                    if (previous != null) {
                        previous.uncaughtException(thread, throwable)
                    } else {
                        // No prior handler: replicate Android's default
                        // termination so the process doesn't hang in a
                        // half-crashed state.
                        android.os.Process.killProcess(android.os.Process.myPid())
                        exitProcess(10)
                    }
                }
            }
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val newChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "dev.ailog/flutter")
        newChannel.setMethodCallHandler(this)
        channel = newChannel
        Ailog.attachChannel(newChannel)
        installCrashHandlerOnce()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> {
                val path = (call.arguments as? Map<*, *>)?.get("logFilePath") as? String
                configuredLogFilePath = path
                result.success(null)
            }
            "emitTestLog" -> {
                Ailog.info(
                    "test message from native (${defaultPlatformLabel()})",
                    context = mapOf("origin" to "emitTestLog"),
                    tags = listOf("native", "smoke-test"),
                )
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel?.let { Ailog.detachChannel(it) }
        channel = null
    }

    private fun defaultPlatformLabel(): String = "Android ${android.os.Build.VERSION.RELEASE}"
}
